tc/doc/tcaboot.1
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "TCABOOT 1"
.TH TCABOOT 1 "2022-11-03" "perl v5.30.0" "TC executables"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tcaboot \- Trusted Collective (TC) authority initialization utility
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBtcaboot\fR \fImulticast_group_number_for_TC_bulletins\fR \fImulticast_group_number_for_TC_records\fR \fInumber_of_authorities_in_collective\fR \fIK\fR \fIR\fR [ \fIdelay\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtcaboot\fR writes a \s-1TC\s0 authority administration command file that initializes
a \s-1TC\s0 authority database.  The file, named \*(L"boot.tcarc\*(R", is written to the
current working directory.  It simply contains two authority configuration
commands that initialize the \s-1TC\s0 authority database for the \s-1TC\s0 application
and then set the initial bulletin compilation time for this authority to
the current ctime plus \fIdelay\fR seconds.  If omitted, \fIdelay\fR defaults
to 5.  The other command-line arguments for \fBtcaboot\fR are discussed in
the descriptions of application initialization commands for \fBtcaadmin\fR;
see the \fBtcarc\fR\|(5) manual page for details.
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
.ie n .IP """0""" 4
.el .IP "``0''" 4
.IX Item "0"
Successful generation of \s-1TC\s0 authority initialization file.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.IP "tcaboot 210 209 6 50 .2" 4
.IX Item "tcaboot 210 209 6 50 .2"
Writes a boot.tcarc file that initializes the local node's \s-1TC\s0 authority
database as indicated and sets the next bulletin compilation time to the
current time plus 5 seconds.
.IP "tcaboot 210 209 6 50 .2 90" 4
.IX Item "tcaboot 210 209 6 50 .2 90"
Writes a boot.tcarc file that initializes the local node's \s-1TC\s0 authority
database as indicated and sets the next bulletin compilation time to the
current time plus 90 seconds.
.SH "FILES"
.IX Header "FILES"
No files apply.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
No environment variables apply.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
The following diagnostics may be issued to the log file:
.IP "Can't open cmd file" 4
.IX Item "Can't open cmd file"
\&\fBtcaboot\fR is unable to create a file named boot.tcarc for the indicated reason,
a system error.
.IP "Can't write to cmd file" 4
.IX Item "Can't write to cmd file"
\&\fBtcaboot\fR is unable to write to boot.tcarc for the indicated reason,
a system error.
.SH "BUGS"
.IX Header "BUGS"
Report bugs to <ion\-dtn\-support@lists.sourceforge.net>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBtcaadmin\fR\|(1), \fBtcarc\fR\|(5)


release-notes.txt
= Release Notes for ION 4.1.1 =

November 19, 2021

%%%%%%%%%%%
= GENERAL =
%%%%%%%%%%%

The ION (interplanetary overlay network) software is a suite of communication
protocol implementations designed to support mission operation communications
across an end-to-end interplanetary network, which might include on-board
(flight) subnets, in-situ planetary or lunar networks, proximity links, deep
space links, and terrestrial internets.  Included in the ION software
distribution are the following packages:

- ici (interplanetary communication infrastructure), a set of libraries
  that provide flight-software-compatible support for functions on which
  the other packages rely, such as dynamic memory management, non-volatile
  storage management, and inter-task communication via shared memory.
  The ici libraries are designed to make the porting of IPN software to
  multiple operating systems - Linux, VxWorks, Solaris, etc. - as easy as
  possible. Ici now includes zco (zero-copy object), a library that 
  minimizes the copying of application data as it is encapsulated in 
  multiple layers of protocol structure while traversing the protocol
  stack.

- bp (bundle protocol), an implementation of the Delay-Tolerant
  Networking (DTN) architecture's Bundle Protocol.

- ltp (licklider transmission protocol), a DTN convergence layer protocol
  for reliable transmission over links characterized by long or highly
  variable delay.

- dgr (datagram retransmission), an alternative implementation of ltp
  that is designed for use over the Internet protocol stack.  dgr
  implements congestion control and is designed for relatively high
  performance.

- bssp (bundle streaming service protocol), a DTN convergence layer protocol
  for reliable transmission of streaming data.

- ams - an implementation of the CCSDS Asynchronous Message Service.

- cfdp - a class-1 (Unacknowledged) implementation of the CCSDS File
  Delivery Protocol.

- bss - a Bundle Streaming Service (BSS) for disruption-tolerant reliable
  data streaming.  BSS supports real-time streaming applications by
  passing the bundle payloads to the associated application for immediate
  display of the most recent data while storing all bundle payloads
  received into a database for user-directed playback.

- dtpc (delay-tolerant payload conditioning), an application framework
  providing TCP-like support for end-to-end retransmission and data delivery
  in transmission order without duplication or omission.

Features included:

- cgr - Contract graph routing: a method of dynamic routing designed for
  space based applications of ION, but still usable for terrestrial
  applications. It computes routes using scheduled communication and deals
  with time-varying network topology.

- brs - Bundle relay service: provides interconnectivity between networks
  that do not allow servers (those behind NAT for example). For more
  information, check man brsscla and man brsccla.

%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.1 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

Beginning with the ION 4.1.1 release, we will no longer be distributing 
two separate .tar.gz files (one that builds with BP version 6 by default and 
one that builds with BP version 7 by default). Instead, we will distribute a
single ION .tar.gz file that builds with BP version 7 by default. If you wish 
to use ION with BP version 6, you can use use the --enable-bpv6 configure 
option. If you are building ION from source on Windows for BP version 6, you
will need to run the command "make bpv6".

After ION version 4, all BP version 6 related changes will be bug fixes
unless explicitly identified as otherwise.

Branches included in this release:

misc-4.1.1-fixes
  * Improve the killm script 
  * Update pilot delay and add rate control for bpdriver
  * Change SourceCustodyRequired to NoCustodyRequested for AMS
  * Add a version of the req-003-multicast test for BP version 7
  * Fix a bug in librams.c that caused AMS petition assert/cancel
    flip-flop
  * Various updates for Makefiles (autoconf and manual)

apl-4.1.1-security-patches
  * Added BPSec Python test suite
  * Added accomodation of NULL terminator in security policy parameters
  * Added checks for NULL pointer references when a BCB targets an 
    extension block
  * Added TEST_POINT macros to BIB and BCB code to log test events

last-pre-git-tweaks
  * Add a warning about potential forwarding loops
  * Fix a typo in biberc documentation

autoconf-updates-MacOS_M1-WSL
  * Fix line ending encoding error in install_macos_sysctl.sh script
  * Add support for building ION on MacOS using the new M1 processor
    * At this time, ION works correctly on the M! processor with MacOS
      version 11.6 (Big Sur) but does not run properly on version
      12.0.1 (Monterey)
  * Add support to configure ION to build on WSL2 but not WSL1
  * ION has also now been tested on RedHat Linux


%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.0 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

ION 4.1.0 is the first release in which the BPv7 implementation is
conformant to the BPv7 protocol specification that will shortly be
published by IETF as a standards-track RFC.

The accompanying BPsec implementation is similarly conformant to
the BPSec specification that will also soon be a standard-track RFC.
NOTE, though, that the TCP convergence-layer protocol implementation
in ION 4.1.0 still conforms to the original TCPCL specification (RFC
7242), NOT to the upcoming TCPCL standards-track RFC; that upgrade
will be supported in a future release.

Another key change in this release is that Aggregate Custody Signaling
is no longer enabled by default.  (Due to some new features in ACS
configuration, when ACS is enabled the bpv6 bpclock daemon will no
longer run successfully unless the ACS database is initialized.
Since not all ION bpv6 users currently initialize ACS in node startup,
disabling ACS by default seems to be the best way forward.)  If you

tc/authority/tcaboot.c
/*
	tcaboot.c:	Trusted Collective authority initialization.
									*/
/*	Copyright (c) 2020, California Institute of Technology.		*/
/*	All rights reserved.						*/
/*	Nodeor: Scott Burleigh, Jet Propulsion Laboratory		*/

#include "tc.h"

#if defined (ION_LWT)
int	tcaboot(int a1, int a2, int a3, int a4, int a5,
		int a6, int a7, int a8, int a9, int a10)
{
	int	bulletinsGroupNbr = a1 ? atoi((char *) a1) : -1;
	int	recordsGroupNbr = a2 ? atoi((char *) a2) : -1;
	int	auth = a3 ? atoi((char *) a3) : 6;
	int	fec_K = a4 ? atoi((char *) a4) : 50;
	doubtle	fec_R = a5 ? atof((char *) a5) : .2;
	int	delay = a6 ? atoi((char *) a6) : 5;
#else
int	main(int argc, char **argv)
{
	int	bulletinsGroupNbr = argc > 1 ? atoi(argv[1]) : -1;
	int	recordsGroupNbr = argc > 2 ? atoi(argv[2]) : -1;
	int	auth = argc > 3 ? atoi(argv[3]) : 6;
	int	fec_K = argc > 4 ? atoi(argv[4]) : 50;
	double	fec_R = argc > 5 ? atof(argv[5]) : .2;
	int	delay = argc > 6 ? atoi(argv[6]) : 5;
#endif
	int	fd;
	time_t	currentTime;
	char	timestamp[TIMESTAMPBUFSZ];
	char	line[256];

	fd = iopen("boot.tcarc", O_WRONLY | O_CREAT | O_TRUNC, 0777);
	if (fd < 0)
	{
		putSysErrmsg("Can't open cmd file", "boot.tcarc");
		return 1;
	}

	isprintf(line, sizeof line, "1 %d %d %d %d %.2f\n",
			bulletinsGroupNbr, recordsGroupNbr, auth, fec_K, fec_R);
	if (write(fd, line, strlen(line)) < 0)
	{
		putSysErrmsg("Can't write to cmd file", line);
		return 1;
	}

	currentTime = getCtime();
	currentTime += delay;
	writeTimestampUTC(currentTime, timestamp);
	isprintf(line, sizeof line, "m compiletime %s\n", timestamp);
	if (write(fd, line, strlen(line)) < 0)
	{
		putSysErrmsg("Can't write to cmd file", line);
		return 1;
	}

	close(fd);
	writeMemo("Stopping tcaboot.");
	return 0;
}
